src/fft/fft.jl

```diff
@@ -0,0 +1,388 @@
+    X = ROCArray(complex(T), p.osz)
+    Y = ROCArray(T, p.sz)
+    xtype = rocfft_transform_type_real_inverse
+    pp = create_plan(xtype, p.sz, T, inplace, p.region)
+    scale = normalization(Y, p.region)
+    ScaledPlan(rROCFFTPlan{complex(T),ROCFFT_INVERSE,inplace,N}(pp..., X, p.sz, xtype, p.region), scale)
+end
+
+function plan_inv(p::rROCFFTPlan{T,ROCFFT_INVERSE,inplace,N}) where {T<:rocfftComplexes,N,inplace}
+    X = ROCArray(real(T), p.osz)
+    xtype = rocfft_transform_type_real_forward
+    pp = create_plan(xtype, p.osz, T, inplace, p.region)
+    scale = normalization(X, p.region)
+    ScaledPlan(rROCFFTPlan{real(T),ROCFFT_FORWARD,inplace,N}(pp..., X, p.sz, xtype, p.region), scale)
+end
+
+## plan execution
+
+function assert_applicable(p::ROCFFTPlan{T}, X::ROCArray{T}) where T
+    (size(X) == p.sz) ||
+        throw(ArgumentError("rocFFT plan applied to wrong-size input"))
+end
+
+function assert_applicable(p::ROCFFTPlan{T,K}, X::ROCArray{T}, Y::ROCArray{Ty}) where {T,K,Ty}
+    assert_applicable(p, X)
+    (size(Y) == p.osz) ||
+        throw(ArgumentError("RocFFT plan applied to wrong-size output"))
+    # type errors should be impossible by dispatch, but just in case:
+    if p.xtype == rocfft_transform_type_real_inverse
+        (Ty == real(T)) ||
+            throw(ArgumentError("Type mismatch for argument Y"))
+    elseif p.xtype == rocfft_transform_type_real_forward
+        (Ty == complex(T)) ||
+            throw(ArgumentError("Type mismatch for argument Y"))
+    else
+        (Ty == T) ||
+            throw(ArgumentError("Type mismatch for argument Y"))
+    end
+end
+
+function unsafe_execute!(plan::cROCFFTPlan{T,K,true,N}, X::ROCArray{T,N}) where {T,K,N}
+    rocfft_execute(plan, [X.handle,], C_NULL, plan.execution_info)
+end
+
+function unsafe_execute!(plan::cROCFFTPlan{T,K,false,N}, X::ROCArray{T,N}, Y::ROCArray{T}) where {T,N,K}
+    Xcopy = copy(X) # since input array can also be modified
+    rocfft_execute(plan, [Xcopy.handle,], [Y.handle,], plan.execution_info)
+end
+
+function unsafe_execute!(plan::rROCFFTPlan{T,ROCFFT_FORWARD,false,N}, X::ROCArray{T,N}, Y::ROCArray{<:rocfftComplexes,N}) where {T<:rocfftReals,N}
+    @assert plan.xtype == rocfft_transform_type_real_forward
+    Xcopy = copy(X)
+    rocfft_execute(plan, [Xcopy.handle,], [Y.handle,], plan.execution_info)
+end
+
+function unsafe_execute!(plan::rROCFFTPlan{T,ROCFFT_INVERSE,false,N}, X::ROCArray{T,N}, Y::ROCArray{<:rocfftReals,N}) where {T<:rocfftComplexes,N}
+    @assert plan.xtype == rocfft_transform_type_real_inverse
+    Xcopy = copy(X)
+    rocfft_execute(plan, [Xcopy.handle,], [Y.handle,], plan.execution_info)
+end
+
+
+function LinearAlgebra.mul!(y::ROCArray{Ty}, p::ROCFFTPlan{T,K,false}, x::ROCArray{T}) where {T,Ty,K}
+    assert_applicable(p, x, y)
+    unsafe_execute!(p, x, y)
+    # 0 is default stream
+    hipStreamSynchronize(Ptr{Cvoid}(UInt64(0)))
```

Comment:
I'm not super confident this will scale well, since there's no guarantee other operations (like those launched as kernels from AMDGPUnative) will use the same queue; we should remember to return to this at some point (not this PR).